internal/catalog/internal/controllers/endpoints/controller.go

```diff
@@ -0,0 +1,331 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package endpoints
+
+import (
+	"context"
+	"sort"
+
+	"github.com/hashicorp/consul/internal/catalog/internal/controllers/workloadhealth"
+	"github.com/hashicorp/consul/internal/catalog/internal/types"
+	"github.com/hashicorp/consul/internal/controller"
+	"github.com/hashicorp/consul/internal/resource"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+const (
+	endpointsMetaManagedBy = "managed-by-controller"
+)
+
+// The WorkloadMapper interface is used to provide an implementation around being able
+// to map a watch even for a Workload resource and translate it to reconciliation requests
+type WorkloadMapper interface {
+	// MapWorkload conforms to the controller.DependencyMapper signature. Given a Workload
+	// resource it should report the resource IDs that have selected the workload.
+	MapWorkload(context.Context, controller.Runtime, *pbresource.Resource) ([]controller.Request, error)
+
+	// TrackIDForSelector should be used to associate the specified WorkloadSelector with
+	// the given resource ID. Future calls to MapWorkload
+	TrackIDForSelector(*pbresource.ID, *pbcatalog.WorkloadSelector)
+
+	// UntrackID should be used to inform the tracker to forget about the specified ID
+	UntrackID(*pbresource.ID)
+}
```

Comment:
One potential other edge case I've been thinking through with this is when the server restarts. In that case, when the server restarts, we will:
1. Start the controller. At this point, we already have some data in the store but the workload mapper is not yet populated because we haven't reconciled anything. 
2. Running the controller will start the watches for all types asynchronously.
3. It may be possible that by the time we need to map the workload we still haven't reconciled any services or service endpoints pointing to it, and so they will not be in the mapper. It would probably be affected by the number of things we need to add to the queue as well.

I think this is similar to the case we chatted about offline earlier today, but I don't think the order of calls in the reconcile will fix it completely because maybe the `Reconcile` that would populate the mapper hasn't been called yet.

Maybe for that case, we need some sort of fall-back mechanism so that we don't miss events in case the mapper is empty (like when there's a cache miss). So if the mapper doesn't return any requests, then maybe we'd do the naive check and look through all services.